timemachine/fe/dummy.py

```diff
@@ -1,151 +1,202 @@
+        for bond_anchor_cc_pairs in product(
+            *[
+                [(bond_anchor, cc) for dummy_atom in cc for bond_anchor in get_bond_anchors(dummy_atom, cc)]
+                for cc in nx.connected_components(induced_g)
+            ]
+        )
+    )
+
+    return dummy_group_assignments
+
+
+def generate_anchored_dummy_group_assignments(
+    mol_a, mol_b, core_atoms_a: Collection[int], core_atoms_b: Collection[int]
+) -> Iterator[Dict[int, Tuple[Optional[int], FrozenSet[int]]]]:
+    """Returns an iterator over candidate anchored dummy group assignments.
+
+    An anchored dummy group assignment is a set of triples (dummy group, j = bond anchor atom, k = angle anchor atom),
+    where dummy atoms are connected to the core only through j, and k is a core neighbor of j where the bond (j, k)
+    exists in both mol_a and mol_b. Note that k may be missing (None) if there are no valid choices.
+
+    See documentation for :py:func:`timemachine.fe.dummy.generate_dummy_group_assignments` for information on how
+    candidate dummy group assignments are generated.
+
+    Example
+    -------
+    >>> mol_a = Chem.MolFromSmiles("CC")
+    >>> mol_b = Chem.MolFromSmiles("c1(C)ccc1")
+    >>> core_atoms_a = [0, 1]
+    >>> core_atoms_b = [2, 0]
+    >>> list(generate_anchored_dummy_group_assignments(mol_a, mol_b, core_atoms_a, core_atoms_b))
+    [{0: (2, frozenset({1})), 2: (0, frozenset({3, 4}))},
+    {0: (2, frozenset({1, 3, 4}))}]
 
-    cc = list(nx.connected_components(induced_g))
+    Parameters
+    ----------
+    mol_a, mol_b: Mol
+        Source and target molecules for alchemical transformation.
+        Dummy atoms are added to mol_a to transform it into a supergraph of mol_b.
 
-    # check that we didn't miss any dummy atoms
-    assert np.sum([len(c) for c in cc]) == (N - len(core))
+    core_atoms_a: collection of int
+        mol_a atoms in the core
 
-    return cc
+    core_atoms_b: collection of int
+        mol_b atoms in the core
 
+    Returns
+    -------
+    iterator of dict
+        each element is a mapping from bond anchor atom to the pair (angle anchor atom, dummy group)
+    """
 
-def canonicalize_bond(ixn):
+    bonds_b = get_romol_bonds(mol_b)
+    bond_graph_b = convert_bond_list_to_nx(bonds_b)
+    dummy_group_assignments = generate_dummy_group_assignments(bond_graph_b, core_atoms_b)
+
+    core_bonds_c = get_core_bonds(mol_a, mol_b, core_atoms_a, core_atoms_b)
+    c_to_b = {c: b for c, b in enumerate(core_atoms_b)}
+    core_bonds_b = frozenset(translate_bonds(core_bonds_c, c_to_b))
+
+    def get_angle_anchors(bond_anchor):
+        valid_angle_anchors = [
+            angle_anchor
+            for angle_anchor in [n for n in bond_graph_b.neighbors(bond_anchor) if n in core_atoms_b]
+            if canonicalize_bond((bond_anchor, angle_anchor)) in core_bonds_b
+        ]
+        return valid_angle_anchors or [None]
+
+    anchored_dummy_group_assignments = (
```

Comment:
nit: Can you add a short comment? The 4 levels of for loops here took me a bit to parse.

Reply:
Added in https://github.com/proteneer/timemachine/pull/1019/commits/8c5087628184388d8b8db2da24d0f6d522e8c853